File: apps/ios/ScribbleFit/Sources/Features/Sync/Domain/SyncScribblesUseCase.swift
import Foundation

public final class SyncScribblesUseCase: Sendable {
    private let scribbleRepository: any ScribbleRepository
    private let engine: any LLMEngine

    public init(scribbleRepository: any ScribbleRepository, engine: any LLMEngine) {
        self.scribbleRepository = scribbleRepository
        self.engine = engine
    }

    public func execute() async {
        for await scribbles in scribbleRepository.getPendingScribbles().values {
            for scribble in scribbles {
                guard case .raw(let id, _, let rawText, _) = scribble else { continue }
                try? await scribbleRepository.updateSyncStatus(id: id, status: .processing)
                let result = await engine.parseWorkout(rawText: rawText)
                if result.status == .success {
                    try? await scribbleRepository.updateSyncStatus(id: id, status: .completed)
                } else {
                    try? await scribbleRepository.updateSyncStatus(id: id, status: .failed)
                }
            }
            break
        }
    }
}

File: apps/android/feature/scribble/domain/src/main/java/com/scribblefit/feature/scribble/domain/SyncScribblesUseCase.kt
package com.scribblefit.feature.scribble.domain

import com.scribblefit.feature.ai.domain.engine.LLMEngine
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory

class SyncScribblesUseCase(
    private val scribbleRepository: ScribbleRepository,
    private val engine: LLMEngine,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    private val logger = LoggerFactory.getLogger("SyncScribblesUseCase")
    suspend operator fun invoke() = withContext(coroutineDispatcher) {
        scribbleRepository.getPendingScribbles().collect { scribbles ->
            for (item in scribbles) {
                val rawText = item.rawText

                engine.parseWorkout(rawText)
                    .fold(
                        onSuccess = { result ->
                            this@SyncScribblesUseCase.scribbleRepository.updateSyncStatus(
                                item.id,
                                SyncStatus.Completed(result)
                            ).fold(
                                onSuccess = {
                                    logger.info("Successfully parsed workout for item ${item.id}")
                                },
                                onFailure = { error ->
                                    logger.error("Error updating sync status for item ${item.id}: ${error.message}")
                                }
                            )
                        },
                        onFailure = { error ->
                            logger.error("Error parsing workout for item ${item.id}: ${error.message}")
                            scribbleRepository.updateSyncStatus(
                                item.id,
                                SyncStatus.Failed
                            )
                        }
                    )
            }
        }
    }
}
